src/Lime.Protocol.UnitTests/Network/ChannelBaseTests.cs

```diff
@@ -311,7 +311,33 @@ public async Task SendMessageAsync_TokenCancelled_DoNotCloseTransport()
             // Assert
             _transport.Verify(t => t.CloseAsync(It.IsAny<CancellationToken>()), Times.Never());
         }
-        
+
+        [Test]
+        [Category("SendMessageAsync")]
+        public async Task SendMessageAsync_LargeEnvelope_DoesNotCompleteTheBuffer()
+        {
+            // Arrange
+            var tcs = new TaskCompletionSource<Envelope>();
+            _transport
+                .Setup(t => t.SendAsync(It.IsAny<Envelope>(), It.IsAny<CancellationToken>()))
+                .Throws(new EnvelopeTooLargeException("too large envelope sent"));
+            var content = Dummy.CreateTextContent();
+            var message = Dummy.CreateMessage(content);
+            var target = GetTarget(SessionState.Established);
+
+            // Act
+            await target.SendMessageAndDelayAsync(message, CancellationToken.None);
+
+            // Assert - should resolve other messages
+            Assert.DoesNotThrowAsync(() => target.SendMessageAndDelayAsync(message, CancellationToken.None), "The send buffer is complete");
```

Comment:
Why are you calling `SendMessageAndDelayAsync` twice?

Reply:
Ah I think I got it. I will wait for you response anyway. But I think the comment could be clearer ^_^

Reply:
The previous code would trigger the undesired behavior at the sencond call of sendMessageAsync. It would enter at the SenderCaller.SendToBufferAsync and stop the channel at line

` if (_envelopeBuffer.Reader.Completion.IsCompleted)
                {
                    throw new InvalidOperationException("The send buffer is complete",
                        _envelopeBuffer.Reader.Completion.Exception?.GetBaseException());
                }`

Reply:
Yea, got it. I think then that the first call should be at the arrange, as you're arranging the scenario required for the test.

Then, the second call should be done on the Act - or at least declared there and stored in a delegate -  as you're doing the act itself that you want to observe the behavior of.

Reply:
did it

Reply:
Just a last bit of suggestion: When we have this case of wanting to check whether or not a method throws an exception - just so that we follow the AAA pattern - we, conventionally, declare the call and store it in a delegate (a `Task`, for instance, or a `Func<T>` when the method isn't asynchronous) and then only call it in the assert (via `await`ing, for example, for `Task`s).

```suggestion
// Act
var sendMessageTask = target.SendMessageAndDelayAsync(message, CancellationToken.None);

// Assert

```
Assert.DoesNotThrowAsync(() => sendMessageTask, "The send buffer is complete");
```